File: examples/php/pwa/schema.php
<?php
/**
 * Schema builder for PHP example
 * See: https://github.com/FranBarInstance/neutralts-docs
 */

require_once __DIR__ . '/constants.php';

// It is important to distinguish between data coming from the user and data
// coming from the application. "CONTEXT" has some security measures such as escaping.
class Schema
{
    protected $schema = [];
    protected $route;

    public function __construct($route)
    {
        $this->route = trim($route, '/\\');
        $this->default();
        $this->populate_context();
        $this->negotiate_language();
        $this->set_theme();
    }

    protected function default()
    {
        $json = file_get_contents(DEFAULT_SCHEMA);
        $this->schema = json_decode($json, true);
        if (!isset($this->schema['data'])) $this->schema['data'] = [];
        if (!isset($this->schema['data']['CONTEXT'])) $this->schema['data']['CONTEXT'] = [];
        $context = &$this->schema['data']['CONTEXT'];
        $context['GET'] = $context['GET'] ?? [];
        $context['POST'] = $context['POST'] ?? [];
        $context['COOKIES'] = $context['COOKIES'] ?? [];
        $context['HEADERS'] = $context['HEADERS'] ?? [];
    }

    protected function populate_context()
    {
        $context = &$this->schema['data']['CONTEXT'];
        $context['ROUTE'] = $this->route;

        // Parse headers using getallheaders() function
        if (function_exists('getallheaders')) {
            $headers = getallheaders();
            foreach ($headers as $key => $value) {
               $context['HEADERS'][$key] = $value;
            }
        } else {
            // Fallback for servers that don't have getallheaders()
            foreach ($_SERVER as $key => $value) {
                if (substr($key, 0, 5) === 'HTTP_') {
                    $headerName = str_replace('_', '-', substr($key, 5));
                    $context['HEADERS'][$headerName] = $value;
                }
            }
        }

        // GET params
        foreach ($_GET as $k => $v) {
            $context['GET'][$k] = $v;
        }

        // POST params
        if ($_SERVER['REQUEST_METHOD'] === 'POST') {
            foreach ($_POST as $k => $v) {
                $context['POST'][$k] = $v;
            }
        }

        // Cookies
        foreach ($_COOKIE as $k => $v) {
            $context['COOKIES'][$k] = $v;
        }

        // Fake session
        $context['SESSION'] = $context['COOKIES']['SESSION'] ?? null;
    }

    protected function negotiate_language()
    {
        $languages = $this->schema['data']['site']['validLanguages'];

        // Use GET, COOKIE or Accept-Language (basic parsing)
        $current = $this->schema['data']['CONTEXT']['GET'][LANG_KEY]
            ?? $this->schema['data']['CONTEXT']['COOKIES'][LANG_KEY]
            ?? $this->_best_match_language($languages)
            ?? '';

        if (!in_array($current, $languages, true)) {
            $current = $languages[0];
        }

        $this->schema['inherit']['locale']['current'] = $current;
    }

    protected function _best_match_language(array $languages)
    {
        // Very small Accept-Language parser: return first matching language
        if (!isset($_SERVER['HTTP_ACCEPT_LANGUAGE'])) return null;
        $accept = $_SERVER['HTTP_ACCEPT_LANGUAGE'];
        $parts = explode(',', $accept);
        foreach ($parts as $p) {
            $lang = strtolower(substr(trim($p), 0, 2));
            foreach ($languages as $valid) {
                if (strtolower(substr($valid, 0, 2)) === $lang) return $valid;
            }
        }
        return null;
    }

    protected function set_theme()
    {
        $this->schema['data']['site']['theme'] =
            $this->schema['data']['CONTEXT']['GET'][THEME_KEY]
            ?? $this->schema['data']['CONTEXT']['COOKIES'][THEME_KEY]
            ?? $this->schema['data']['site']['validThemes'][0];
    }

    public function get()
    {
        return $this->schema;
    }
}

File: examples/go/pwa/schema.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
)

type Schema struct {
	schema map[string]interface{}
	route  string
}

func NewSchema(r *http.Request, route string) *Schema {
	s := &Schema{
		route:  strings.Trim(route, "/\\"),
		schema: make(map[string]interface{}),
	}
	s.defaultSchema()
	s.populateContext(r)
	s.negotiateLanguage(r)
	s.setTheme(r)
	return s
}

func (s *Schema) defaultSchema() {
	fullPath := DEFAULT_SCHEMA
	data, err := os.ReadFile(fullPath)
	if err != nil {
		fmt.Printf("Error reading schema: %v\n", err)
		return
	}
	if err := json.Unmarshal(data, &s.schema); err != nil {
		fmt.Printf("Error unmarshaling schema: %v\n", err)
		return
	}
	if _, ok := s.schema["data"]; !ok {
		s.schema["data"] = make(map[string]interface{})
	}
	if _, ok := s.schema["data"].(map[string]interface{})["CONTEXT"]; !ok {
		s.schema["data"].(map[string]interface{})["CONTEXT"] = make(map[string]interface{})
	}
	s.context()["GET"] = make(map[string]interface{})
	s.context()["POST"] = make(map[string]interface{})
	s.context()["COOKIES"] = make(map[string]interface{})
	s.context()["HEADERS"] = make(map[string]interface{})
}

func (s *Schema) schemaData() map[string]interface{} {
	return s.schema["data"].(map[string]interface{})
}

func (s *Schema) context() map[string]interface{} {
	return s.schemaData()["CONTEXT"].(map[string]interface{})
}

func (s *Schema) populateContext(r *http.Request) {
	s.context()["ROUTE"] = s.route

	// Headers
	headers := make(map[string]interface{})
	for name, values := range r.Header {
		headerName := strings.ToUpper(name)
		if len(values) > 0 {
			headers[headerName] = values[0]
		}
	}
	s.context()["HEADERS"] = headers
	headers["HOST"] = r.Host

	// GET
	for key, values := range r.URL.Query() {
		if len(values) > 0 {
			s.context()["GET"].(map[string]interface{})[key] = values[0]
		}
	}

	// POST
	if r.Method == "POST" {
		if err := r.ParseForm(); err == nil {
			for key, values := range r.PostForm {
				if len(values) > 0 {
					s.context()["POST"].(map[string]interface{})[key] = values[0]
				}
			}
		}
	}

	// Cookies
	cookies := make(map[string]interface{})
	for _, cookie := range r.Cookies() {
		cookies[cookie.Name] = cookie.Value
	}
	s.context()["COOKIES"] = cookies

	// Session
	session, exists := cookies["SESSION"]
	if exists {
		s.context()["SESSION"] = session
	} else {
		s.context()["SESSION"] = nil
	}
}

func (s *Schema) negotiateLanguage(r *http.Request) {
	siteData, ok := s.schema["data"].(map[string]interface{})["site"].(map[string]interface{})
	if !ok {
		return
	}
	validLanguages, ok := siteData["validLanguages"].([]interface{})
	if !ok {
		return
	}
	langList := make([]string, len(validLanguages))
	for i, l := range validLanguages {
		langList[i] = l.(string)
	}

	current := s.getParam(LANG_KEY, "GET", "COOKIES")
	if current == "" {
		// Simple Accept-Language match
		acceptLang := r.Header.Get("Accept-Language")
		if acceptLang != "" {
			parts := strings.Split(acceptLang, ",")
			for _, part := range parts {
				langPart := strings.SplitN(strings.TrimSpace(part), ";", 2)[0]
				lang := strings.ToLower(langPart[:2]) // First two letters
				for _, valid := range langList {
					if strings.HasPrefix(strings.ToLower(valid), lang) {
						current = valid
						break
					}
				}
				if current != "" {
					break
				}
			}
		}
	}
	if current == "" || !s.contains(langList, current) {
		if len(langList) > 0 {
			current = langList[0]
		}
	}
	if _, ok := s.schema["inherit"]; !ok {
		s.schema["inherit"] = make(map[string]interface{})
	}
	if _, ok := s.schema["inherit"].(map[string]interface{})["locale"]; !ok {
		s.schema["inherit"].(map[string]interface{})["locale"] = make(map[string]interface{})
	}
	s.schema["inherit"].(map[string]interface{})["locale"].(map[string]interface{})["current"] = current
}

func (s *Schema) getParam(key, source1, source2 string) string {
	ctx := s.context()
	switch source1 {
	case "GET":
		if get, ok := ctx["GET"].(map[string]interface{}); ok {
			if val, ok := get[key]; ok {
				return val.(string)
			}
		}
	case "COOKIES":
		if cookies, ok := ctx["COOKIES"].(map[string]interface{}); ok {
			if val, ok := cookies[key]; ok {
				return val.(string)
			}
		}
	}
	// source2 same logic
	switch source2 {
	case "GET":
		if get, ok := ctx["GET"].(map[string]interface{}); ok {
			if val, ok := get[key]; ok {
				return val.(string)
			}
		}
	case "COOKIES":
		if cookies, ok := ctx["COOKIES"].(map[string]interface{}); ok {
			if val, ok := cookies[key]; ok {
				return val.(string)
			}
		}
	}
	return ""
}

func (s *Schema) contains(slice []string, item string) bool {
	for _, a := range slice {
		if a == item {
			return true
		}
	}
	return false
}

func (s *Schema) setTheme(r *http.Request) {
	theme := s.getParam(THEME_KEY, "GET", "COOKIES")
	if theme == "" {
		siteData, ok := s.schema["data"].(map[string]interface{})["site"].(map[string]interface{})
		if ok {
			validThemes, ok := siteData["validThemes"].([]interface{})
			if ok && len(validThemes) > 0 {
				theme = validThemes[0].(string)
			}
		}
	}
	schemaData := s.schema["data"].(map[string]interface{})
	if _, ok := schemaData["site"]; !ok {
		schemaData["site"] = make(map[string]interface{})
	}
	s.schema["data"].(map[string]interface{})["site"].(map[string]interface{})["theme"] = theme
}

func (s *Schema) Get() map[string]interface{} {
	return s.schema
}
